### ISSUE TITLE
sentence needs to be improved

### ISSUE THREAD

**Issue by PinoTsao** [0xAX/linux-insides#645]
In "Kernel booting process. Part 4.", section "Reload the segments if needed", one sentence should be improved:

>The value assigned to it is an offset relative to the offset of the segment

==>

>The value assigned to it is an offset relative to the segment

**Comment by tkyymmt** [comment]
Would you mind sending PR?

**Comment by PinoTsao** [comment]
yes, but I need to learn how to sending a PR, please wait for a while:)

**Comment by PinoTsao** [comment]
> After comparison of the LOAD_PHYSICAL_ADDR and value of the ebx register, we add the offset from the startup_32 where to decompress the compressed kernel image. If the CONFIG_RELOCATABLE option is not enabled during kernel configuration, we just put the default address where to load kernel and add z_extract_offset to it.

To be honest, I think most content of it is wrong. It is just a calculation of physical address where to place the decompressed kernel, if we don't have CONFIG_RELOCATABLE, it is the LOAD_PHYSICAL_ADDR, or else, it will be the address of %ebp(protect mode kernel loaded address) aligned to BP_kernel_alignment, while no less than LOAD_PHYSICAL_ADDR

**Comment by PinoTsao** [comment]
>The compressed kernel image should be moved to the end of the decompression buffer to simplify calculations where kernel will be located later. For this:

Although there is no obvious buffer ending address in the code, but I feel "moved against to the end xxx" is more accurate, "against" has more meaning that the compressed kernel wouldn't exceed the buffer.  And I don't quite understand why the latter part exists, how about:

>The compressed kernel image will be moved against the end of the decompression buffer for safety in-place decompression.

**Comment by PinoTsao** [comment]
> setting 5 bit in it and loading it again into cr4

==>

> setting the 5th bit in it and loading it again into cr4

**Comment by PinoTsao** [comment]
> > After comparison of the LOAD_PHYSICAL_ADDR and value of the ebx register, we add the offset from the startup_32 where to decompress the compressed kernel image. If the CONFIG_RELOCATABLE option is not enabled during kernel configuration, we just put the default address where to load kernel and add z_extract_offset to it.
> 
> To be honest, I think most content of it is wrong, for example, I don't see any reference of z_extract_offset in current code here, maybe in older time it is different.
> 
> It is just a calculation of physical address where to place the decompressed kernel, if we don't have CONFIG_RELOCATABLE, it is the LOAD_PHYSICAL_ADDR, or else, it will be the address of %ebp(protect mode kernel loaded address) aligned to BP_kernel_alignment, while no less than LOAD_PHYSICAL_ADDR.

I find the [commit](https://lore.kernel.org/patchwork/patch/674095/) which drop the symbol: z_extract_offset

**Comment by WarpspeedSCP** [comment]
I'm currently doing some proofreading work where I try to correct spelling, grammar and formatting mistakes, and I've found another thing which seems a bit off...

on line 549,

> Here we put the `MSR_EFER` flag (which is defined in [arch/x86/include/asm/msr-index.h](https://github.com/torvalds/linux/blob/v4.16/arch/x86/include/asm/msr-index.h)) in the `ecx` register and call `rdmsr` instruction which reads the [MSR](http://en.wikipedia.org/wiki/Model-specific_register) register. After `rdmsr` executes, we will have the resulting data in `edx:eax` which depends on the `ecx` value. We check the `EFER_LME` bit with the `btsl` instruction and write data from `eax` to the `MSR` register with the `wrmsr` instruction.

here, shouldn't `eax` in 

> We check the `EFER_LME` bit with the `btsl` instruction and write data from `eax` to the `MSR` register with the `wrmsr` instruction.

be `edx:eax`?

**Comment by mudongliang** [comment]
Right. `wrmsr` instruction writes the contents of registers EDX:EAX into the 64-bit model specific register (MSR) specified in the ECX register.

**Comment by 0xAX** [comment]
I will close the issue since this chapter was rewritten. @WarpspeedSCP thank you for this issue, despite it is probably late, please feel free to re-open.
